realtek: make RTL8226/RTL8221B work on RTL930x

plappermaul opened the pull request:
This series backports two RTL8226 phy patches, that went upstream in the last weeks. With that:

- RTL8226 will be detected by upstream driver
- RTL8226 will switch its serdes depending on the connection speed (SGMII or 2500base-x)
- The ZyXEL XGS1210-10 DTS needs to be converted to 2500base-x
- The RTL8221B soft_reset() downstream patch gets a fix.

This small series comes with a caveat. The RTL8226 ports give new/other/more issues than before. We cannot fix them all in one series. But now there is only the way forward - no time to look back. Luckily we already have other PRs that will take care about the remaining bits.

- https://github.com/openwrt/openwrt/pull/19518
- https://github.com/openwrt/openwrt/pull/19834

plappermaul committed:
generic: backport upstream v6.18 Realtek PHY patch

34167f1a0 net: phy: realtek: convert RTL8226-CG to c45 only

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19843
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul committed:
generic: backport upstream v6.18 Realtek PHY patch

3a752e678 net: phy: realtek: enable serdes option mode for RTL8226-CG

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19843
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul committed:
realtek: XGS1210-12: convert RTL8226 PHYs to 2500base-x

We reached the point of no return. Upstream has gained the final
bits for the RTL8226 PHYs. That means.

- RTL8226 MAC side behaves like RTL8221(B)
- It's serdes no longer uses proprietary HSGMII (2.5G SGMII)
- Instead it dynamically switches between SGMII and 2500base-x

This (partly) solves one of the central henn/egg problems of the
Realtek target. To change the MAC/PHY interface mode both sides
need to have all bits in place to do so. But where to start if
so much needs to be done?

Now the PHY side has created facts and it mitigates a lot of
problems. All downstream HSGMII patches and coding can be dropped
in the future.

For now only adapt the only DTS that still maps PHYs to HSGMII.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19843
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul commented:
No idea as of now. Feel free to have a look here https://github.com/plappermaul/realtek-doc/tree/main/sources/rtk-dms1250/src/hal/phy

plappermaul commented:
@bevanweiss While preparing this series I thought about our c22/c45 forum discussion. Current assumption: our downstream phy patches are not mode consistent. Would be nice if you can give this a try.

- Make a build on top of this series
- It includes a new c45 capable soft reset function
- Switch DTS to c45 for the RTL8221B phys
- Modify realtek_main.c as follows:

```
		.match_phy_device = rtl8221b_vb_cg_c45_match_phy_device,
		.name           = "RTL8221B-VB-CG 2.5Gbps PHY (C45)",
		.config_intr	= rtl8221b_config_intr,
		.handle_interrupt = rtl8221b_handle_interrupt,
		.soft_reset     = genphy_soft_reset, --> CHANGE TO rtl822x_c45_soft_reset
		.probe		= rtl822x_probe,
		.config_init    = rtl822xb_config_init_war,
		.get_rate_matching = rtl822xb_get_rate_matching,
		.get_features   = rtl822x_c45_get_features,
		.config_aneg    = rtl822x_c45_config_aneg,
		.read_status    = rtl822xb_c45_read_status,
		.suspend        = genphy_c45_pma_suspend,
		.resume         = rtlgen_c45_resume,
```

abajk commented:
> No idea as of now. Feel free to have a look here

Looking at these sources, it is hard to determine what is supported by each generation of Realtek PHYs.
Do you have access to  hardware? Can you test this [patch](https://github.com/abajk/linux/commit/0c688532c6f3f808b162316ddf90681307b9eeeb) by installing lm-sensors and executing 'sensors' on the console??

janh commented:
@abajk I tried that before, and it didn't work for the RTL8226 in my XGS1010-12.

bevanweiss commented:
> Can you test this [patch](https://github.com/abajk/linux/commit/0c688532c6f3f808b162316ddf90681307b9eeeb) by installing lm-sensors and executing 'sensors' on the console??

I think there would be more value in just using mdio-tools to look at the applicable registers (from RTL8221b datasheet)...
MMD 31.0xBD84   (Temperature)
MMD 31.0xB54C   (Temperature limits)

bevanweiss commented:
@plappermaul I see one of your changes is hsgmii => 2500base-x in a dts, and I note that upstream doesn't have a definition for a HSGMII.  In situations where a PHY might only support HSGMII and not 2500base-x (for a 2.5gbps speed), how would that currently be captured?
It would be nice if HSGMII could just be purged from OpenWRT entirely, but I'm sure there will be some PHYs out there that support only HSGMII.. unsure if that could just be SGMII with some speed applied (to get the 3.125g interface rate)

plappermaul commented:
Branch looks good. Sad to hear. So extra thanks for taking the time for this vague idea. The only chance to get closer to it, would be to add a dump_stack() directly above each "return -EIO" or "err = -EIO" statement in the mdio driver. If possible give it a try. Otherwise the issue has to wait ...

bevanweiss commented:
I did just try having some code that toggled the c22/c45 mode as required based on which method was called... and it seemed to do 'good things'...
```
[    2.915438] rtl83xx_enable_phy_polling:          f110101
[    2.921703] rtl83xx-switch switch@1b000000: led_set0 has 3 LEDs configured
[    2.929589] rtl83xx-switch switch@1b000000: configuring for fixed/internal link mode
[    2.938741] rtl83xx-switch switch@1b000000: Link is Up - 10Gbps/Full - flow control off
[    3.035727] rtl83xx-switch switch@1b000000 lan1 (uninitialized): PHY [mdio-bus:00] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[    3.155796] rtl83xx-switch switch@1b000000 lan2 (uninitialized): PHY [mdio-bus:08] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[    3.275744] rtl83xx-switch switch@1b000000 lan3 (uninitialized): PHY [mdio-bus:10] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[    3.395757] rtl83xx-switch switch@1b000000 lan4 (uninitialized): PHY [mdio-bus:14] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[    3.586757] RTL8221B-VB-CG 2.5Gbps PHY (C45) mdio-bus:18: rtl822xb_config_init failed: -145
[    3.596137] RTL8221B-VB-CG 2.5Gbps PHY (C45) mdio-bus:18: SerDes setup timed out, retrying
[    3.706157] RTL8221B-VB-CG 2.5Gbps PHY (C45) mdio-bus:18: rtl822xb_config_init failed: -145
[    3.715725] rtl83xx-switch switch@1b000000 lan5 (uninitialized): failed to connect to PHY: -ETIMEDOUT
[    3.726107] rtl83xx-switch switch@1b000000 lan5 (uninitialized): error -145 setting up PHY for tree 0, switch 0, port 24
[    3.835794] rtl83xx-switch switch@1b000000 lan6 (uninitialized): PHY [mdio-bus:19] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[    3.955805] rtl83xx-switch switch@1b000000 lan7 (uninitialized): PHY [mdio-bus:1a] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[    4.075871] rtl83xx-switch switch@1b000000 lan8 (uninitialized): PHY [mdio-bus:1b] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[    4.091349] rtl838x-eth 1b00a300.ethernet eth0: entered promiscuous mode
[    4.099086] DSA: tree 0 setup
[    4.102496] LINK state irq: 23
[    4.106027] In rtl83xx_setup_qos
[    4.109853] rtl930x_dbgfs_init called
[    4.114020] rtl83xx_fib_event_work_do: FIB4 default rule failed
```

The -ETIMEDOUT might be caused by some of the other things I was trying...
https://github.com/openwrt/openwrt/compare/main...bevanweiss:openwrt:rtl93xx_mdio_c22_c45_dyn_change

My first hope was that there was just some timing glitch around the MMD flag in the register, just based on how the RTK SDK seems to set these bits all individually, which is uncommon for them, they normally seem to just dump a random magic number into a register at once.  But I didn't notice any change with splitting out the EXEC from the other flag setting (although it's possible I've messed something up here).
Toggling the state of the c22/c45 flag seemed to do things better.  I think the bus state should be cached and compared to avoid the IO write if not necessary though... and there's likely a cleaner way to actually implement the function (I'm open to suggestions around this).

janh commented:
Tested on XGS1010-12, works as expected (as I already had similar patches applied locally).

Remaining issues are the same ones I already mentioned elsewhere:
- Configuring EEE gives an error. However, it still works despite this (reduction in power usage is the same as with vendor firmware).
  ```
  root@OpenWrt:~# ethtool --set-eee lan10 eee on
  [   98.750387] Enabled EEE for port 25
  netlink error: I/O error
  ```
- Port LEDs stay on even when link is down. This only happens in SGMII mode.

The RTL8226 PHYs also only work with `rtk network on` in the bootloader. #19834 doesn't change this.

About the temperature sensor: When adding a patch similar to the one suggested by @abajk, hwmons are registered, but the temperature is always reported as 0. The actual registers in the PHYs look like this:
```
root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 24:31 0xBD84
0x0000
root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 24:31 0xB54C
0xdb18
root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 25:31 0xBD84
0x0000
root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 25:31 0xB54C
0xdb18
```

plappermaul commented:
Enablement bits for thermal sensor a different for gen1. https://github.com/plappermaul/realtek-doc/blob/54589ff0afa70045abfdc71a3133aa55c76257bc/sources/rtk-dms1250/src/hal/phy/nic_rtl8226/nic_rtl8226.c#L1736

janh commented:
@plappermaul I tried to enable the thermal sensor by writing to the registers using the `mdio` tool, but that doesn't change anything.

abajk commented:
> About the temperature sensor: When adding a patch similar to the one suggested by @abajk, hwmons are registered, but the temperature is always reported as 0. The actual registers in the PHYs look like this:
> 
> ```
> root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 24:31 0xBD84
> 0x0000
> root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 24:31 0xB54C
> 0xdb18
> root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 25:31 0xBD84
> 0x0000
> root@OpenWrt:~# mdio rtldsa_mdio-0 mmd 25:31 0xB54C
> 0xdb18
> ```

Thanks for the Info. Can you still read registers 2 and 3 containing OUI?

bevanweiss commented:
> Why else would upstream have a c22-only driver sibling?

There are lots of legacy devices that don't support c45.. so I understand that where practicable some newer PHYs that want c45 can still work fine if they only get c22... then there's the whole c45 over c22.. which it seems has plenty of weird bugs and gotchas with different vendor devices (*cough* Realtek)..

plappermaul commented:
> I did just try having some code that toggled the c22/c45 mode as required based on which method was called... and it seemed to do 'good things'...

Then simply add a pr_info("... port ...") and dump_stack() to the c45->c22 switching statement and we know when it is going wrong.

plappermaul committed:
generic: fix c45 soft reset for RTL8221B

The addition of the soft_reset() function to the RTL8221B PHYs
missed to take care of C22/C45 standalone PHY versions. Especially
on RTL930x switch devices with these PHY the reset fails for the
C45 operation mode. This comes from the fact that the mdio bus
disables C22 read/writes when being set to C45.

Upstream has gained a proper C45 reset function. Use it for the
C45 PHY models.

Fixes: 7e3284eef7f ("generic: use genphy_soft_reset for RealTek 2.5G PHYs")

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19843
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul commented:
> root@OpenWrt:~# ethtool --set-eee lan10 eee on
> [   98.750387] Enabled EEE for port 25
> netlink error: I/O error

Memo for myself. Fails because of C22 presence bit. Thus phy_restart_aneg() takes c22 path.
```
phy_restart_aneg(struct phy_device *phydev)
{
	int ret;

	if (phydev->is_c45 && !(phydev->c45_ids.devices_in_package & BIT(0))) --> C22 = BIT 0 set
		ret = genphy_c45_restart_aneg(phydev);
	else
		ret = genphy_restart_aneg(phydev);

	return ret;
}
```

plappermaul commented:
To much in parallel these days. Rebased and fixed.

abajk commented:
> About the temperature sensor: When adding a patch similar to the one suggested by @abajk, hwmons are registered, but the temperature is always reported as 0

The first generation of Realtek 2.5G PHY apparently does not have a temperature sensor.
First gen:
* 0x001cc800- RTL8226 -> Unknown. Testing required.
* 0x001cc838 - RTL8226-CG -> No temperature sensor. Tested by @janh.

Second gen:
* 0x001cc840 - RTL8226B/RTL8221B -> Unknown. Needs testing,
* 0x001cc848 - RTL8226B-CG/RTL8221B-CG -> Unknown. Needs testing.

Third gen:
* 0x001cc849 - RTL8221B-VB-CG -> Sensor available. Tested by @abajk.
* 0x001cc84a - RTL8221B-VM-CG -> Sensor available.

Maybe the second generation has a built-in sensor. Does anyone have access to hardware with the second generation Realtek PHY?

plappermaul commented:
@abajk convinced some non-human-intelligence that some documentation is free for all. Its answer regarding thermal sensors was:

Enabling the Sensor (Step-by-Step)
By default, the thermal ADC is often powered down to save energy. You must manually enable the thermal bias and the ADC clock.

For RTL8226 / RTL8226B:

- Access MMD 31, Register 0x6601.
- Set Bit 12 (THS_EN): This is the main thermal sensor enable bit.
- Set Bit 11 (ADC_EN): Powers up the ADC that converts the analog diode voltage to a digital value.
- Recommended Write: Read the current value and OR it with 0x1800.

For RTL8221B:
- Access MMD 31, Register 0x6601.
- Set Bit 12 (THS_EN): Main enable.
- Check Bit 15 (LDO_OFF): Ensure this is 0 (Normal operation).
- Note: The 8221B often requires a "Trigger" bit to refresh the ADC value. If the value stays frozen, you may need to toggle Bit 10 of the same register to force a conversion.

Not really verified.
